Experiments

adrian-gierakowski opened the pull request.

jcarnegie committed:
More Effect integration with RTK

jcarnegie committed:
Working version of createApiFromEffectLayer

adrian-gierakowski commented:
@jcarnegie created this PR so that we can have an async discussion about the code

adrian-gierakowski commented on vite-template-redux/src/app/effect.ts:
```diff
@@ -0,0 +1,107 @@
+import type { ThunkExtraArgument } from '../app/store';
+
+// Helper type that extracts the success type from an Effect
+type ExtractSuccessType<T> = T extends Effect.Effect<infer A, any, any> ? A : never;
```
I guess you intentionally not using [Effect.Effect.Success](https://github.com/Effect-TS/effect/blob/22905cf5addfb1ff3d2a6135c52036be958ae911/packages/effect/src/Effect.ts#L240) so that non Effect types could be passed into it?

jcarnegie replied:
No, it was a holdover from some code borrowed from another context. I'm updating it to use Effect.Effect.Success

jcarnegie committed:
Added type check tests for createApiFromEffectLayer

adrian-gierakowski committed:
better type safety for default test iml serviceMap

adrian-gierakowski reviewed (commented):
looks good!

I played with adding some more type safety to test serviceMap, and pushed some changes [here](https://github.com/rhinofi/redux-toolkit-play/pull/2). Let me know what you think

adrian-gierakowski commented on vite-template-redux/src/app/effect.ts:
```diff
@@ -0,0 +1,107 @@
+}[keyof S];
+
+// Main function that creates an RTK Query API from an Effect Layer
+export const createApiFromEffectLayer = <
```
`createApiFromEffectService` or `FromEffectTag` would probably be more appropriate

jcarnegie replied:
Let's go with FromEffectTag. At one point I was passing in a layer, but found ways to do what we want with just the Tag.

adrian-gierakowski commented on vite-template-redux/src/app/effect.ts:
```diff
@@ -0,0 +1,107 @@
+        return yield* service[methodKey](args);
+      }).pipe(Effect.either);
+      const { extra } = api as { extra: ThunkExtraArgument };
+      const resultEither = await extra.runtime.runPromise(effect);
```
to allow RTK query to interrupt the effect:

```suggestion
      const resultEither = await extra.runtime.runPromise(effect, { signal: api.signal });
```

jcarnegie replied:
Ohhh...nice catch.

adrian-gierakowski commented on vite-template-redux/src/app/effect.ts:
```diff
@@ -0,0 +1,107 @@
+    const method = (serviceTag as any)[methodKey] as SI[typeof methodKey];
+
+    return async (args: Parameters<typeof method>[0], api: BaseQueryApi) => {
+      const effect = Effect.gen(function* () {
+        const service = yield* serviceTag;
+        return yield* service[methodKey](args);
+      }).pipe(Effect.either);
```
We've got some use cases in which we need to access redux state in the query handlers (for example to get JWT token for currently logged in user), so would be good to do something like:

```ts
export class ReduxState extends Context.Tag("ReduxState")<
  ReduxState,
  unknown
>() {}
```

and then

```suggestion
      const effect = Effect.gen(function* () {
        const service = yield* serviceTag;
        return yield* service[methodKey](args);
      }).pipe(
        Effect.provideService(ReduxState, api.getState()), 
        Effect.either
      );
```

We can leave state typed as `unknown` (since that's what RTK does) and leave it to the code in the effect to cast or parse it as necessary.

jcarnegie replied:
Yup, makes sense. I can see other use cases for injecting app state into effects.

adrian-gierakowski commented on vite-template-redux/src/app/effect.ts:
```diff
@@ -0,0 +1,107 @@
+
+// Main function that creates an RTK Query API from an Effect Layer
+export const createApiFromEffectLayer = <
+  S, 
+  SI extends { [K in keyof SI]: (...args: any[]) => Effect.Effect<any, any, any> }
+>(
+  // Takes a service tag (Context.Tag) and configuration options
+  serviceTag: Context.Tag<S, SI>,
```
Probably worth restricting the R of the effects returned by services methods and the serviceTag

```ts
  S extends ManagedRuntime.ManagedRuntime.Context<ThunkExtraArgument['runtime']>,
  SI extends { [K in keyof SI]: (...args: any[]) => Effect.Effect<any, any, S> }
>(
  // Takes a service tag (Context.Tag) and configuration options
  serviceTag: Context.Tag<S, SI>,
```

adrian-gierakowski replied:
we can also throw in the `ReduxState` (see other comment)

```ts
  S extends ManagedRuntime.ManagedRuntime.Context<ThunkExtraArgument['runtime']> | ReduxState,
```

jcarnegie replied:
This will break the test in effect.test-d.ts. Let me think of a way where we can restrict R but keep all the tests passing.

adrian-gierakowski replied:
I merged your branch into mine and pushed a tentative fix for the issue you raised [here](https://github.com/rhinofi/redux-toolkit-play/pull/2/commits/1c30652c3b32bbd609038c2c7ca27286ebdf7e16)

It also allows any service to have runtime dependency on any other  service within the runtime (just in case) as well as ReduxState.

jcarnegie replied:
Ah, a factory that takes an explicit type param. Clever. Great, let me dig into this more tomorrow or Sunday.

jcarnegie committed:
Incorporated feedback from Adrian: use Effect.Effect.Success, rename to createApiFromEffectTag, effect interruption, inject app state into effects

adrian-gierakowski committed:
Merge branch 'experiments' into ag/experiments

adrian-gierakowski committed:
more type safety in createApiFromEffectLayer

allow creating instances of createApiFromEffectLayer linked to different
RutimeServices type

adrian-gierakowski committed:
rename: createApiFromEffectLayer => createApiFromEffectTag

adrian-gierakowski committed:
format with: cd vite-template-redux && dprint fmt 'src/**'

adrian-gierakowski committed:
add .vscode/settings.recommended.json to enable format on save with dprint

adrian-gierakowski committed:
Merge pull request #3 from rhinofi/ag/format

ag/format

adrian-gierakowski committed:
Merge pull request #2 from rhinofi/ag/experiments

More Effect integration with RTK

jcarnegie committed:
Updated createApiFromEffectTagFactory with Adrian's latest updates and copied types from rtk query so we don't have to patch the package

adrian-gierakowski committed:
minor cleanup

adrian-gierakowski committed:
fix ErrorType

adrian-gierakowski committed:
add effect/platform

adrian-gierakowski committed:
add HttpApiClient example

adrian-gierakowski committed:
createApiFromEffectTag: allow configuring a subset of methods

adrian-gierakowski committed:
fix src/app/effect.test-d.ts

adrian-gierakowski committed:
fix test-utils after adding FromHttpApiClient

adrian-gierakowski committed:
tsconfig: add exactOptionalPropertyTypes: true

jcarnegie committed:
Added flattenHttpApiClient to transform an HttpApiClient into something we can feed into createApiFromEffectTag

adrian-gierakowski committed:
fix type errors due to enabling tsconfig.exactOptionalPropertyTypes

adrian-gierakowski committed:
update pnpm lock

adrian-gierakowski committed:
Merge pull request #5 from rhinofi/ag/allow-configuring-subset-of-methods-final

ag/allow configuring subset of methods final

adrian-gierakowski committed:
add tsconfig.verbatimModuleSyntax=true, cleanup imports and dprint fmt

adrian-gierakowski committed:
add tsconfig.declaration=true and fix type errors

adrian-gierakowski commented:
Merging to master to make it easier for others to browse the code. We can create a new PR for further changes
